compiler/src/dotty/tools/dotc/util/HashSet.scala

```diff
@@ -67,22 +69,22 @@ class HashSet[T](initialCapacity: Int = 8, capacityMultiple: Int = 2) extends Mu
   /** Turn hashcode `x` into a table index */
   protected def index(x: Int): Int = x & (table.length - 1)
 
-  protected def currentTable: Array[AnyRef] = table
+  protected def currentTable: Array[AnyRef | Null] = table
 
   protected def firstIndex(x: T) = if isDense then 0 else index(hash(x))
   protected def nextIndex(idx: Int) =
     Stats.record(statsItem("miss"))
     index(idx + 1)
 
-  protected def entryAt(idx: Int) = table(idx).asInstanceOf[T]
-  protected def setEntry(idx: Int, x: T) = table(idx) = x.asInstanceOf[AnyRef]
+  protected def entryAt(idx: Int): T | Null = table(idx).asInstanceOf[T | Null]
+  protected def setEntry(idx: Int, x: T) = table(idx) = x.asInstanceOf[AnyRef | Null]
 
   def lookup(x: T): T | Null =
     Stats.record(statsItem("lookup"))
     var idx = firstIndex(x)
-    var e = entryAt(idx)
+    var e: T | Null = entryAt(idx)
     while e != null do
-      if isEqual(e, x) then return e
+      if isEqual(e.asInstanceOf[T], x) then return e.asInstanceOf[T]
```

Comment:
.uncheckedNN. We should avoid `asInstanceOf` for not-null assertions.

Reply:
Done